osu.Game.Rulesets.Osu/Mods/OsuModClassic.cs

```diff
@@ -59,12 +64,32 @@ public void ApplyToDrawableHitObject(DrawableHitObject obj)
             {
                 case DrawableSliderHead head:
                     head.TrackFollowCircle = !NoSliderHeadMovement.Value;
+                    if (FadeHitCircleEarly.Value)
+                        applyEarlyFading(head);
                     break;
 
                 case DrawableSliderTail tail:
                     tail.SamplePlaysOnlyOnHit = !AlwaysPlayTailSample.Value;
                     break;
+
+                case DrawableHitCircle circle:
+                    if (FadeHitCircleEarly.Value)
+                        applyEarlyFading(circle);
+                    break;
             }
         }
+
+        private void applyEarlyFading(DrawableHitCircle circle)
+        {
+            circle.ApplyCustomUpdateState += (o, _) =>
+            {
+                using (o.BeginAbsoluteSequence(o.StateUpdateTime))
+                {
+                    double mehWindow = o.HitObject.HitWindows.WindowFor(HitResult.Meh);
+                    double lateMissFadeTime = mehWindow / 4 + 15;
+                    o.Delay(mehWindow - lateMissFadeTime).FadeOut(lateMissFadeTime);
```

Comment:
stable code:

```csharp
            Transformation fadeout = hidden ?
                new Transformation(TransformationType.Fade, 1, 0, StartTime - (int)(LocalPreEmpt * 0.6), StartTime - (int)(LocalPreEmpt * 0.3)) { TagNumeric = ARMED } :
                new Transformation(TransformationType.Fade, 1, 0, StartTime + hitObjectManager.HitWindow100, StartTime + hitObjectManager.HitWindow50) { TagNumeric = ARMED };
```

So two things to check / match:

- Make sure this change doesn't apply when hidden mod is turned on (may depend on the order of mod application, so needs care)
- The implementation you have can be improved slightly: the fade starts precisely at hit100 window and ends at hit50 window.

Reply:
side note, but please don't resolve review conversations yourself. we do this after coming back and checking they have been applied satisfactorily.